### ISSUE TITLE
Pretty printing with big uncertainties

### ISSUE THREAD

**Issue by bjodah** [bjodah/chempy#39]
```
>>> from chempy.printing import number_to_scientific_unicode
>>> number_to_scientific_unicode(0.05, 5.5)
...
ValueError: unsupported format character '-' (0x2d) at index 2
```

**Comment by bjodah** [comment]
Using +/- (\pm) should also be possible by passing a kwarg `plusminus=True` to any of the `number_to_scientific_*` functions.

**Comment by Solero93** [comment]
I tried to fix this bug to start contributing to this repo, but I couldn't get my head around how this scientific representation with number, standard error and precision works.

I think the code for formatting can be simplified, but I'd like to ask for an explanation of how this notation works. I even looked at https://stackoverflow.com/questions/6671053/python-pretty-print-errorbars but it is still very confusing to me.

Could you elaborate with examples?

**Comment by Solero93** [comment]
Okay, that is where my problems begin. Why not for example:

```python
>>> number_to_scientific_unicode(0.05, 5.5)
'0.5(55)·10⁻¹
```

? I understand the Python code, I debugged it, I found the line, wrote a test that reproduces the bug, but I don't know what output I should expect, because I don't fully understand what this function should be doing... I see the examples, but not the reasoning behind them. Basically, in general, if a value and a standard error is given, what do you do to arrive to this representation?

Because I see that the idea is something along the lines of changing the standard error to a whole number and then scaling the value so that everything makes sense. Does this notation have a name? Because I couldn't find anything about it. 

The +- notation is much easier to do, because I understand how that notation is generated

**Comment by bjodah** [comment]
Sorry, I misunderstood your question. I guess this is a matter of canonicalization, i.e. we just need a reasonable set of rules which give us one unambiguous representation. The rule which I've been following rather unconsciously is:

 - only one single digit goes before the decimal place, and that digit should not be zero (which raises the question how `number_to_scientific_unicode(0, 1.23)` should print, but let's defer that for now).

I could also see merit in making the exponent a multiple of three (since we're so used to thinking in k, M, G, T, ... etc.).

I don't know if the notation with parenthesis has a name, here are a two reputable resources I found while looking into that:

- https://physics.nist.gov/cgi-bin/cuu/Info/Constants/definitions.html
- https://journals.aps.org/prl/authors/general-notation-terminology

I haven't seen any python package which can output strings in this format ([uncertainties can parse it](https://pythonhosted.org/uncertainties/user_guide.html#creating-numbers-with-uncertainties)), the latex package [siunitx](http://ftp.dante.de/tex-archive/macros/latex/exptl/siunitx/siunitx.pdf) does use it by default.

**Comment by Solero93** [comment]
Yes, the main problem here is that this conversion is not unique. As for your rule of thumb, I am seeing that most test cases don't follow it:

```python
assert _float_str_w_uncert(-5739, 16.34, 2) == '-5739(16)'
assert _float_str_w_uncert(-5739, 16.9, 2) == '-5739(17)'
assert _float_str_w_uncert(0.0123, 0.00169, 2) == '0.0123(17)'
assert _float_str_w_uncert(0.01234, 0.00169, 2) == '0.0123(17)'
assert _float_str_w_uncert(0.01234, 0.0016501, 2) == '0.0123(17)'
assert _float_str_w_uncert(0.01234, 0.0016501, 1) == '0.012(2)'
assert _float_str_w_uncert(-9.99752e15, 349e10, 2) == '-9.9975(35)e15'
assert _float_str_w_uncert(-9.99752e5, 349, 2) == '-999750(350)'
assert _float_str_w_uncert(-9.99752e5, 349, 3) == '-999752(349)'
assert _float_str_w_uncert(315, 17.9e-4, 2) == '315.0000(18)'
```

This is why I wanted to clarify what the function should be doing.

I have an idea to standardize this, it makes more or less sense with the test cases and it should work in all cases. Let me know your opinion about it.

The idea relies on scaling the standard error to a whole number and adapting the value to match the scaling. 

So, for example, if the standard error is 0.1234 and I have precision 3, then the number in parenthesis would be 123 (because 0.1234 as a whole number would be scaled to 1234 but we just use the first 3 digits). After that, we would scale the value to match the scaling of the standard error. So if the value is 15 and standard error is 0.1234 with precision 3, it would become 1500(12)e-2 (because we adapted 15 to the scale of the standard error of (12), i.e. multiplied the value by 10^2, so that the 12 makes sense inside the notation). We added e-2 to the end because that's how we would go back from 1500 to 15. 

For this case, the result would be

```python
>>> number_to_scientific_unicode(0, 1.23)
'0(123)e-2'
>>> number_to_scientific_unicode(0.5, 5.5)
'5(55)e-1'
>>> number_to_scientific_unicode(0.05, 5.5)
'0.5(55)e-1'
```

I hope I explained myself about the idea. I left out cases like having to add padding 0s to the decimals, but that's just part of "adapting to the standard error's scale". I can elaborate on any edge case if you have doubts.

**Comment by bjodah** [comment]
Ah, you're right. I like the idea, it feels consistent. Feel free to go ahead and change it to use this format instead.

**Comment by Solero93** [comment]
Awesome! I'll do the pull request as soon as I have it

**Comment by bjodah** [comment]
Great! Do you intend to add the "plusminus" option as well?

**Comment by Solero93** [comment]
I don't think so, because usually adding a boolean parameter to decide what
a function does is sign of a bad design. Don't get me wrong, sometimes it
is necessary to decide things inside a function, but if the logic of the
whole function is based on "if random_boolean: do this, else: do other
thing", then that's usually a red flag.

I'd like to first stabilize this respresentation and check that it works
properly.

I think the plusminus should be a separate issue, because it is a
completely different thing. This is a bug and plusminus is a new feature.

Are you okay with that?

There are many ways to add new kinds of representations such as
implementing something similar to the Strategy pattern
https://en.m.wikipedia.org/wiki/Strategy_pattern. But that can be evaluated
inside the other issue.

**Comment by Solero93** [comment]
I tried to rewrite the test cases to match my idea, but I'm not sure if this is what the function should be doing:

Original test:

```python
assert _float_str_w_uncert(-5739, 16.34, 2) == '-5739(16)'
assert _float_str_w_uncert(-5739, 16.9, 2) == '-5739(17)'
assert _float_str_w_uncert(0.0123, 0.00169, 2) == '0.0123(17)'
assert _float_str_w_uncert(0.01234, 0.00169, 2) == '0.0123(17)'
assert _float_str_w_uncert(0.01234, 0.0016501, 2) == '0.0123(17)'
assert _float_str_w_uncert(0.01234, 0.0016501, 1) == '0.012(2)'
assert _float_str_w_uncert(-9.99752e15, 349e10, 2) == '-9.9975(35)e15'
assert _float_str_w_uncert(-9.99752e5, 349, 2) == '-999750(350)'
assert _float_str_w_uncert(-9.99752e5, 349, 3) == '-999752(349)'
assert _float_str_w_uncert(315, 17.9e-4, 2) == '315.0000(18)'
```

Changed test (and added new ones):

```python
assert _float_str_w_uncert(-5739, 16.34, 2) == '-5739(16)'
assert _float_str_w_uncert(-5739, 16.9, 2) == '-5739(17)'
assert _float_str_w_uncert(0.0123, 0.00169, 2) == '123(17)e-4'
assert _float_str_w_uncert(0.01234, 0.00169, 2) == '123(17)e-4'
assert _float_str_w_uncert(0.01234, 0.0016501, 2) == '123(17)e-4'
assert _float_str_w_uncert(0.01234, 0.0016501, 1) == '12(2)e-3'
assert _float_str_w_uncert(-9.99752e15, 349e10, 2) == '-99975(35)e11'
assert _float_str_w_uncert(-9.99752e5, 349, 2) == '-99975(35)e1'
assert _float_str_w_uncert(-9.99752e5, 349, 3) == '-999752(349)'
assert _float_str_w_uncert(315, 17.9e-4, 2) == '3150000(18)e-4'
assert _float_str_w_uncert(0, 1.23, 2) == '00(12)e-1'
assert _float_str_w_uncert(0, 1.23, 3) == '000(123)e-2'
assert _float_str_w_uncert(0.5, 5.5, 2) == '05(55)e-1'
assert _float_str_w_uncert(0.05, 5.5, 2) == '005(550)e-2'
```

Is this new representation okay with you?

**Comment by bjodah** [comment]
I would actually prefer to have a `scilimits` option analogous to [matplotlib](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.ticklabel_format.html#matplotlib.axes.Axes.ticklabel_format), we could default it to e.g. `(-4, 4)` (which is what python uses for the `%g` format string).

Here is a motivating example of why it's preferable—[Planck temperature](https://physics.nist.gov/cgi-bin/cuu/Category?view=pdf&Universal.x=116&Universal.y=16):
```
assert _float_str_w_uncert(1.416808e32, 33e26) == 1.416808(33)e32
```
It's immediately obvious that the constant is ~10**32 -- no need to digits and adding the number of digits to the exponent to figure out what order of magnitude the number is...
